src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,393 @@
+    enc = !EVP_CIPHER_CTX_encrypting(ctx);
+#endif
+
+    /* Trying to encrypt */
+    ret = Esys_EncryptDecrypt2( eactx.ectx,
+                                keyHandle,
+                                ESYS_TR_PASSWORD,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                in_data,
+                                enc,
+                                mode,
+                                &iv_in,
+                                &out_data,
+                                &iv_out );
+    if (ret == TPM2_RC_COMMAND_CODE) {
+        DBG("Esys_EncryptDecrypt2 : FAILED\n");
+        ret = Esys_EncryptDecrypt( eactx.ectx,
+                                   keyHandle,
+                                   ESYS_TR_PASSWORD,
+                                   ESYS_TR_NONE,
+                                   ESYS_TR_NONE,
+                                   enc,
+                                   mode,
+                                   &iv_in,
+                                   in_data,
+                                   &out_data,
+                                   &iv_out );
+        if(ret == TPM2_RC_SUCCESS) {
+            DBG("Esys_EncryptDecrypt  : SUCCESS\n");
+        }
+        else {
+            DBG("Esys_EncryptDecrypt  : FAILED\n");
+        }
+    }
+    ERRchktss(tpm2_do_cipher, ret, goto error);
+
+    /* Copy out_data : TPM2B_MAX_BUFFER to unsigned char* */
+    memcpy(out, out_data->buffer, out_data->size);
+    out[out_data->size] = '\0';
+
+    /* Close TPM session */
+    if (keyHandle != ESYS_TR_NONE) {
+        if (tpm2Data->privatetype == KEY_TYPE_HANDLE) {
+            Esys_TR_Close(eactx.ectx, &keyHandle);
+        } else {
+            Esys_FlushContext(eactx.ectx, keyHandle);
+        }
+    }
+    esys_auxctx_free(&eactx);
+
+    return out_data->size;
+
+error :
+    /* Close TPM session */
+    if (keyHandle != ESYS_TR_NONE) {
+        if (tpm2Data->privatetype == KEY_TYPE_HANDLE) {
+            Esys_TR_Close(eactx.ectx, &keyHandle);
+        } else {
+            Esys_FlushContext(eactx.ectx, keyHandle);
+        }
+    }
+    esys_auxctx_free(&eactx);
+
+    if (tpm2Data)
+        OPENSSL_free(tpm2Data);
+    return 0;
+}
+
+static int
+tpm2_cipher_cleanup(EVP_CIPHER_CTX *ctx)
+{
+    (void)(ctx);
+    DBG("Cleaning up ...\n");
+    return 1;
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+static EVP_CIPHER tpm2_aes_256_ofb =
+{
+    NID_aes_256_ofb128,                 // ID
+    TPM2_MAX_SYM_BLOCK_SIZE,            // Block size
+    TPM2_MAX_SYM_KEY_BYTES,             // Key length
+    TPM2_MAX_SYM_BLOCK_SIZE,            // IV length
+    EVP_CIPH_OFB_MODE,                  // Flags
+    tpm2_cipher_init_key,               // Init key
+    tpm2_do_cipher,                     // Encrypt/Decrypt
+    tpm2_cipher_cleanup,                // Cleanup
+    sizeof(TPM2_DATA),                  // Context size
```

Comment:
I don't think this is correct. The man-page says `EVP_CIPHER_meth_set_impl_ctx_size() sets the size of the EVP_CIPHER's implementation context so that it can be automatically allocated.` but we are malloc'ing it ourselves. But I also don't know what this is really for and how we should use it.

Reply:
To be honest, I don't fully understand this stuff neither.

I based my code by taking the [gost engine](https://github.com/gost-engine/engine) as example. In there code, we can see that there are doing a `EVP_CIPHER_meth_set_impl_ctx_size(_hidden_magma_cbc, sizeof(struct ossl_gost_cipher_ctx))` [here](https://github.com/gost-engine/engine/blob/master/gost_crypt.c#L248) and then `struct ossl_gost_cipher_ctx *c = EVP_CIPHER_CTX_get_cipher_data(ctx);` [here](https://github.com/gost-engine/engine/blob/master/gost_crypt.c#L461).

So maybe I should put a 0 value here, I tried and it does not seem to change anything. I can change it if you want ?